design/service-apis-implementation.md

```diff
@@ -0,0 +1,258 @@
+- Gateway requests a point where traffic can be translated to Services within the cluster, using an internal Listener construct.
+- Routes describe how traffic coming via the Gateway maps to the Services.
+
+In Contour, we've previously solved a lot of the same problems with HTTPProxy (and IngressRoute before it).
+That functionality can be described in the Service APIs by HTTPRoutes and TLSRoutes, as they describe Layer 7 ingress, as Contour does today.
+
+Other types of routes include TCPRoutes and UDPRoutes, which are intended for Layer 4 load balancers. Implementations may also define their own Route objects, at the cost of interoperability.
+
+In terms of its stated goals, Contour is aiming at being an ingress controller - that is, a Layer 7 proxy with some api gateway functions.
+Currently, Contour provides "TCP Proxying" that allows the forwarding of TLS streams based on the SNI information, which is precisely what the Service APIs TLSRoute object is for.
+If Project Contour (the organisation) does add support for TCP and UDP forwarding, it will not be in the `projectcontour/contour` repo, but will be a separate repo.
+
+This design is intended to cover the initial `v1alpha1` release of the Service APIs. We will aim to implement the core featureset of the APIs at that point.
+We will then work with the upstream community on features that Contour and HTTPProxy currently support, but the Service APIs do not, and how best to represent those features in the APIs.
+So, there are some features that are currently gaps for Contour (wildcard domain names, and exact path matching seem like big ones),
+and some that Contour supports that the Service APIs do not (webscokets, configurable timeouts, header replacement, external auth, rate limiting, and so on).
+
+## Goals
+- Define a data model and implementation for Contour's service-apis support, covering the v1alpha1 version of the service-apis.
+- Layer 7 support only, which means HTTPRoutes, TLSRoutes only.
+
+## Non Goals
+- No TCPRoute or UDPRoute support, that is, no support for arbitrary TCP/UDP forwarding.
+
+
+## High-Level Design
+
+### Service APIs implementation surface area
+
+Contour's support for the Service APIs will include support for the HTTPRoute and TLSRoute objects only, not any other type of Route - this includes no TCPRoute or UDPRoute support.
+Contour is a layer 7 ingress controller, and the layer 4 load balancing implies by TCPRoute and UDPRoute is out of scope for this tool (that is, `projectcontour/contour`).
+Project Contour (the organisation) may investigate a Service-APIs based layer 4 solution in the future, but that effort will not be in `projectcontour/contour`.
+
+### Contour Gateway model
+
+Contour considers a Gateway to describe a single Envoy deployment, deployed and managed by something outside of itself.
+Contour expects to be supplied with the full name (name and namespace) of a single Gateway, which it will watch and update the status of.
```

Comment:
Do you see this as new flags to Contour or placed into the config file?

Reply:
I think config file for this one, I'll update.